specs/PushNotifications/PushNotifications-spec.md

```diff
@@ -0,0 +1,438 @@
+
+## Seperating Activator Registration flow from Channel Request flow
+We decided to have the following 2 APIs to be seperate calls instead of a single combined API call:
+```cpp
+PushNotificationChannelManager::RegisterPushNotificationActivator(details)
+PushNotificationChannelManager::CreatePushNotificationChannel(remoteIdentifier)
+```
+Mainly for 2 reasons:
+* The app developer is expected to Register an activator for every WinMain app launch. Combining the channel request API with the registration call would force the developer to keep the client channel in sync with the App Service more frequently (both for foreground and background launch) which can cause potential synchronization bugs. The preference is for developers to request new channels only on Foreground launches triggered by the user.
+* It isn't required that developers Register a Push Activator for Visual Toast operations. In the case of Visual Toasts, payloads are directed to the Shell and not to the App.
+
+## Handling Push ChannelFailures
+PushNotification channel request calls are expected to fail since they go over the wire. We don't want the developer to deal with the complexities of retryable failures and critical failures. So we expose in-progress states and internal error codes for developers to action on.
+
+For example, we have 2 different Progress states, one indicating basic progress and another indicating retry progress.
+```cpp
+if (args.Status() == ChannelStatus::InProgress)
+{
+    // This is basically a noop since it isn't really an error state
+    printf("The first channel request is still in progress! \n");
+}
+else if (args.Status() == ChannelStatus::InProgressRetry)
+{
+    LOG_HR_MSG(args.ExtendedError(), "The channel request is in back-off retry mode because of a retryable error! Expect delays in acquiring it.");
+}
+```
+
+Similarly, operations are expected to fail or succeed on completion.
+```cpp
+if (result.Status() == ChannelStatus::CompletedSuccess)
+{
+    auto channelUri = result.Channel().Uri();
+    auto channelExpiry = result.Channel().ExpirationTime();
+
+    // Persist the channelUri and Expiry in the App Service
+}
+else if (result.Status() == ChannelStatus::CompletedFailure)
+{
+    LOG_HR_MSG(result.ExtendedError(), "We hit a critical non-retryable error with channel request!");
+}
+```
+
+## Manifest Registration
+For MSIX, the COM activator GUID and the exe path need to be registered in the manifest. The launch args would need to be pre set to a well-known string that defines Push Triggers. 
+```xml
+<Extensions>
+  <com:Extension Category="windows.comServer">
+    <com:ComServer>
+      <com:ExeServer Executable="SampleBackgroundApp\SampleBackgroundApp.exe" DisplayName="SampleBackgroundApp" Arguments="-ReunionPushServer">
+        <com:Class Id="{GUIDEntryPoint}" DisplayName="Reunion Push" />
+      </com:ExeServer>
+    </com:ComServer>
+  </com:Extension>
+</Extensions>
+```
+
+## UWP InProc activator
```

Comment:
Is the implementation of in-proc activation for UWP going to be part of the Phase 1?

Reply:
Hmm thats a great question. I would prioritize inproc first and then out-of-proc. I will add an OutOfProc activator too to this design to make things clear.

Reply:
Yes. It's just a simple wrapper for now